lib/key/utils.js

```diff
@@ -1,51 +1,93 @@
+
+/**
+ * Returns the preferred symmetric/aead/compression algorithm for a set of keys
+ * @param {symmetric|aead|compression} type - Type of preference to return
+ * @param {Array<Key>} [keys] - Set of keys
+ * @param {Date} [date] - Use the given date for verification instead of the current time
+ * @param {Array} [userIDs] - User IDs
+ * @param {Object} [config] - Full configuration, defaults to openpgp.config
+ * @returns {Promise<module:enums.symmetric|aead|compression>} Preferred algorithm
+ * @async
+ */
+export async function getPreferredAlgo(type, keys = [], date = new Date(), userIDs = []) {
+    const defaultAlgo = {
+        // these are all must-implement in rfc4880bis
+        symmetric: enums.symmetric.aes128,
+        aead: enums.aead.eax,
+        compression: enums.compression.uncompressed
+    }[type];
+    const preferredSenderAlgo = {
+        symmetric: enums.symmetric.aes256,
+        aead: enums.aead.eax,
+        compression: enums.compression.uncompressed
+    }[type];
+    const prefPropertyName = {
+        symmetric: 'preferredSymmetricAlgorithms',
+        aead: 'preferredAEADAlgorithms',
+        compression: 'preferredCompressionAlgorithms'
+    }[type];
+
+    // if preferredSenderAlgo appears in the prefs of all recipients, we pick it
+    // otherwise we use the default algo
+    // if no keys are available, preferredSenderAlgo is returned
+    const senderAlgoSupport = await Promise.all(
+        keys.map(async (key, i) => {
+            const primaryUser = await key.getPrimaryUser(date, userIDs[i]);
+            const recipientPrefs = primaryUser.selfCertification[prefPropertyName];
+            return !!recipientPrefs && recipientPrefs.indexOf(preferredSenderAlgo) >= 0;
+        })
+    );
+    return senderAlgoSupport.every(Boolean) ? preferredSenderAlgo : defaultAlgo;
+}
 
 // returns promise for generated RSA public and encrypted private keys
-export function generateKey({ passphrase, date = serverTime(), offset = DEFAULT_OFFSET, ...rest }) {
+export async function generateKey({ passphrase, date = serverTime(), offset = DEFAULT_OFFSET, ...rest }) {
     if (!passphrase) {
         throw new Error('passphrase required');
     }
     const offsetDate = new Date(date.getTime() + offset);
-    return openpgp.generateKey({ passphrase, date: offsetDate, ...rest });
+    return openpgpGenerateKey({ passphrase, date: offsetDate, ...rest });
 }
 
-export function generateSessionKey(algorithm) {
-    return openpgp.crypto.generateSessionKey(algorithm);
+export function generateSessionKey(encryptionKey) {
+    return openpgpGenerateSessionKey({ encryptionKeys: [encryptionKey] });
 }
 
 export async function getPreferredAlgorithm(keys, date = serverTime()) {
-    return openpgp.enums.read(openpgp.enums.symmetric, await openpgp.key.getPreferredAlgo('symmetric', keys, date));
+    return enums.read(enums.symmetric, await getPreferredAlgo('symmetric', keys, date));
```

Comment:
I think we can remove this (and `getPreferredAlgo`) and have the apps call `generateSessionKeys` directly (no other usage so far)